### PR TITLE
Update Handshakers to use NodeKey

### PR DESCRIPTION
**Pull request by rain-on**
## PR description
Update the Handhshaking classes to use an injected NodeKey object for ECIES encryption.
This has necessitated an update of NodeKey to support the creation of Key and ECIES agreement creation.

## Fixed Issue(s)
N/A

### CODE REVIEW

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/connections/netty/HandshakeHandlerOutbound.java [hunk 1]
```diff
@@ -53,7 +54,8 @@ public HandshakeHandlerOutbound(
         connectionFuture,
         connectionEventDispatcher,
         metricsSystem);
-    handshaker.prepareInitiator(kp, SECP256K1.PublicKey.create(peer.getId()));
+    handshaker.prepareInitiator(
```
Comment (usmansaleem) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/connections/netty/HandshakeHandlerOutbound.java:57]: Does it make sense to move the second parameter `SECP256K1.PublicKey.create(peer.getId())` as potentially a static method in NodeKey as well?

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/handshake/ecies/ECIESHandshaker.java [hunk 2]
```diff
@@ -101,14 +103,14 @@ public void prepareInitiator(final SECP256K1.KeyPair ourKeypair, final PublicKey
...
+  public void prepareResponder(final NodeKey nodeKey) {
     checkState(
         status.compareAndSet(
             Handshaker.HandshakeStatus.UNINITIALIZED, Handshaker.HandshakeStatus.IN_PROGRESS),
         "handshake was already prepared");
 
     this.initiator = false;
-    this.identityKeyPair = ourKeypair;
+    this.nodeKey = nodeKey;
     this.ephKeyPair = SECP256K1.KeyPair.generate();
```
Comment (usmansaleem) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/handshake/ecies/ECIESHandshaker.java:114]: what about `ephKeyPair`?

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/handshake/ecies/EncryptedMessage.java [hunk 3]
```diff
@@ -57,19 +58,19 @@ public static Bytes decryptMsg(final Bytes msgBytes, final SECP256K1.PrivateKey
...
    * Decrypts the ciphertext using our private key.
    *
    * @param msgBytes The ciphertext.
-   * @param ourKey Our private key.
+   * @param nodeKey Abstraction of this nodes private key & associated crypto operations
    * @return The plaintext.
    * @throws InvalidCipherTextException Thrown if decryption failed.
    */
-  public static Bytes decryptMsgEIP8(final Bytes msgBytes, final SECP256K1.PrivateKey ourKey)
+  public static Bytes decryptMsgEIP8(final Bytes msgBytes, final NodeKey nodeKey)
```
Comment (usmansaleem) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/handshake/ecies/EncryptedMessage.java:73]: just wondering, should we name argument as 'ourNodeKey' ?

File: crypto/src/main/java/org/hyperledger/besu/crypto/BouncyCastleNodeKey.java [hunk 4]
```diff
@@ -41,4 +51,23 @@ public Signature sign(final Bytes32 dataHash) {
...
+    final ECDomainParameters dp = SECP256K1.CURVE;
+
+    final CipherParameters pubParam = new ECPublicKeyParameters(ephPubKey.asEcPoint(), dp);
+    final CipherParameters privParam =
+        new ECPrivateKeyParameters(nodeKeys.getPrivateKey().getD(), dp);
+
+    final BasicAgreement agree = new ECDHBasicAgreement();
+    agree.init(privParam);
+    final BigInteger z = agree.calculateAgreement(pubParam);
+    return Bytes.wrap(BigIntegers.asUnsignedByteArray(agree.getFieldSize(), z));
```
Comment (jframe) [inline crypto/src/main/java/org/hyperledger/besu/crypto/BouncyCastleNodeKey.java:71]: Would be nicer if both of the key agreement functions were in SECP256k1 class

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/connections/netty/HandshakeHandlerOutbound.java (hunk 1 above)
Comment (rain-on) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/connections/netty/HandshakeHandlerOutbound.java:57]: Thinking not - NodeKey is all about protecting this node's identity/private key - the Peer in this question is someone remote.
I.e. only operations for _our_ node key need to go via the NodeKey interface

File: crypto/src/main/java/org/hyperledger/besu/crypto/BouncyCastleNodeKey.java (hunk 4 above)
Comment (rain-on) [reply crypto/src/main/java/org/hyperledger/besu/crypto/BouncyCastleNodeKey.java:71]: done

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/handshake/Handshaker.java [hunk 5]
```diff
@@ -76,11 +77,11 @@
    * <p>This method must throw an {@link IllegalStateException} exception if the handshake had
    * already been prepared before, no matter if under the initiator or the responder role.
    *
-   * @param ourKeypair The keypair for our node identity.
+   * @param nodeKey An object which represents our identity
    * @param theirPubKey The public key of the node we're handshaking with.
    * @throws IllegalStateException Indicates that preparation had already occured.
    */
-  void prepareInitiator(SECP256K1.KeyPair ourKeypair, SECP256K1.PublicKey theirPubKey);
+  void prepareInitiator(final NodeKey nodeKey, SECP256K1.PublicKey theirPubKey);
```
[the remark this answers is not in the document]
Comment (rain-on) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/handshake/Handshaker.java:84]: done

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/handshake/ecies/ECIESEncryptionEngine.java [hunk 6]
```diff
@@ -82,41 +75,30 @@
   private final byte[] iv;
 
   private ECIESEncryptionEngine(
-      final CipherParameters pubParam,
-      final CipherParameters privParam,
-      final SECP256K1.PublicKey ephPubKey,
-      final byte[] iv) {
+      final Bytes Z, final SECP256K1.PublicKey ephPubKey, final byte[] iv) {
```
[the remark this answers is not in the document]
Comment (rain-on) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/handshake/ecies/ECIESEncryptionEngine.java:78]: changed - but I wonder what it REALLY is .... changed to agreedSecret

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/handshake/ecies/ECIESHandshaker.java (hunk 2 above)
Comment (rain-on) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/handshake/ecies/ECIESHandshaker.java:114]: This is the ephemeral key used for inter-node communications and is rolling - so no need to protect it alongside our nodekey

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/handshake/ecies/EncryptedMessage.java (hunk 3 above)
Comment (rain-on) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/handshake/ecies/EncryptedMessage.java:73]: I'd prefer not to. For 2 reasons:
Thinking "nodeKey" is explicit enough to indicate ownership (on some levels).
Secondly, I'd hate for this class to imply the owner of the key which is doing the decryption - i.e. it simply needs a key to do the decryption - ownership of the key makes no difference as to the operation being conducted.

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/handshake/ecies/InitiatorHandshakeMessageV1.java [hunk 7]
```diff
@@ -95,11 +95,10 @@ public static InitiatorHandshakeMessageV1 create(
    * Decodes this message.
    *
    * @param bytes The raw bytes.
-   * @param keyPair Our keypair to calculat ECDH key agreements.
+   * @param nodeKey Our keypair to calculat ECDH key agreements.
```
[the remark this answers is not in the document]
Comment (rain-on) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/handshake/ecies/InitiatorHandshakeMessageV1.java:98]: done

File: crypto/src/main/java/org/hyperledger/besu/crypto/BouncyCastleNodeKey.java [hunk 8]
```diff
@@ -41,4 +51,23 @@ public Signature sign(final Bytes32 dataHash) {
...
     return nodeKeys.getPublicKey();
   }
+
+  @Override
+  public Bytes32 calculateKeyAgreement(final PublicKey partyKey) {
+    return SECP256K1.calculateKeyAgreement(nodeKeys.getPrivateKey(), partyKey);
+  }
+
+  @Override
+  public Bytes calculateECIESAgreement(final SECP256K1.PublicKey ephPubKey) {
```
[the remark this answers is not in the document]
Comment (rain-on) [reply crypto/src/main/java/org/hyperledger/besu/crypto/BouncyCastleNodeKey.java:61]: done

File: crypto/src/main/java/org/hyperledger/besu/crypto/NodeKey.java [hunk 9]
```diff
@@ -17,11 +17,16 @@
...
+import org.apache.tuweni.bytes.Bytes;
 import org.apache.tuweni.bytes.Bytes32;
 
 public interface NodeKey {
 
   Signature sign(Bytes32 dataHash);
 
   PublicKey getPublicKey();
+
+  Bytes32 calculateKeyAgreement(final PublicKey partyKey);
```
[the remark this answers is not in the document]
Comment (rain-on) [reply crypto/src/main/java/org/hyperledger/besu/crypto/NodeKey.java:29]: done

[the remark this answers is not in the document]
Comment (rain-on) [reply crypto/src/main/java/org/hyperledger/besu/crypto/NodeKey.java:29]: done

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/handshake/Handshaker.java [hunk 10]
```diff
@@ -89,10 +90,10 @@
    * <p>This method must throw an {@link IllegalStateException} exception if the handshake had
    * already been prepared before, whether with the initiator or the responder role.
    *
-   * @param ourKeypair The keypair for our node identity.
+   * @param nodeKey An object which represents our identity
    * @throws IllegalStateException Indicates that preparation had already occured.
    */
-  void prepareResponder(SECP256K1.KeyPair ourKeypair);
+  void prepareResponder(final NodeKey nodeKey);
```
[the remark this answers is not in the document]
Comment (rain-on) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/handshake/Handshaker.java:96]: done

**Pull request merged**